File: backend/app/schemas/detection.py
from typing import List, Literal, Optional

from pydantic import BaseModel

from app.schemas.violation import ViolationReport, ZoneViolation


class BoundingBox(BaseModel):
    x1: float
    y1: float
    x2: float
    y2: float


class Detection(BaseModel):
    id: int
    label: str
    category: Literal["compliant", "violation"]
    confidence: float
    bbox: BoundingBox
    color: str


class EquipmentStatus(BaseModel):
    label: str
    status: Literal["compliant", "violation"]
    confidence: Optional[float] = None
    bbox: Optional[BoundingBox] = None


class PersonResult(BaseModel):
    person_id: int
    track_id: Optional[int] = None
    bbox: BoundingBox
    confidence: float
    equipment: List[EquipmentStatus]
    compliant: bool


class Summary(BaseModel):
    total_persons: int
    compliant: int
    violations: int
    inference_ms: float


class DetectionResponse(BaseModel):
    detections: List[Detection]
    persons: List[PersonResult]
    summary: Summary


class VideoSummary(BaseModel):
    video_name: str
    total_frames: int
    processed_frames: int
    fps: float
    duration_seconds: float
    unique_violations: int
    candidate_violations: int = 0
    inference_ms: float


class TrackingOverlayFrame(BaseModel):
    frame_index: int
    time_seconds: float
    track_id: int | None
    person_id: int | None
    bbox: BoundingBox
    confidence: float
    compliant: bool
    missing_equipment: list[str]
    status: Literal["compliant", "violation", "unknown"]
    zone_id: int | None = None
    zone_name: str | None = None
    zone_type: Literal["RESTRICTED", "WALKWAY"] | None = None


class TrackingOverlay(BaseModel):
    fps: float
    stride: int
    frame_width: int | None
    frame_height: int | None
    frames: list[TrackingOverlayFrame]


class VideoProcessingResponse(BaseModel):
    summary: VideoSummary
    reports: List[ViolationReport]
    zone_violations: List[ZoneViolation] = []
    tracking_overlay: TrackingOverlay | None = None

File: frontend/src/types/detection.ts
import { ZoneViolation } from "./zone";

export interface BoundingBox {
  x1: number;
  y1: number;
  x2: number;
  y2: number;
}

export type Category = "compliant" | "violation";

export interface Detection {
  id: number;
  label: string;
  category: Category;
  confidence: number;
  bbox: BoundingBox;
  color: string;
}

export interface EquipmentStatus {
  label: string;
  status: Category;
  confidence?: number;
  bbox?: BoundingBox;
}

export interface PersonResult {
  person_id: number;
  track_id?: number;
  bbox: BoundingBox;
  confidence: number;
  role: "worker" | "janitor" | null;
  uniform_type: "vest" | "cleaning_coverall" | null;
  equipment: EquipmentStatus[];
  compliant: boolean;
}

export interface Summary {
  total_persons: number;
  compliant: number;
  violations: number;
  inference_ms: number;
}

export interface DetectionResponse {
  detections: Detection[];
  persons: PersonResult[];
  summary: Summary;
}

export interface ViolationReport {
  id: number;
  timestamp: string;
  violation_type: string;
  details: string;
  snapshot_url?: string;
  video_name?: string;
  frame_index?: number;
  track_id?: number;
}

export interface VideoSummary {
  video_name: string;
  total_frames: number;
  processed_frames: number;
  fps: number;
  duration_seconds: number;
  unique_violations: number;
  candidate_violations: number;
  inference_ms: number;
}

export interface TrackingOverlayFrame {
  frame_index: number;
  time_seconds: number;
  track_id?: number;
  person_id?: number;
  bbox: BoundingBox;
  confidence: number;
  role?: "worker" | "janitor" | null;
  uniform_type?: "vest" | "cleaning_coverall" | null;
  compliant: boolean;
  missing_equipment: string[];
  zone_id?: number;
  camera_zone_view_id?: number;
  physical_zone_id?: number;
  zone_name?: string;
  zone_type?: "RESTRICTED" | "WALKWAY" | "SLIPPERY";
  status: Category | "unknown";
}

export interface TrackingOverlay {
  fps: number;
  stride: number;
  frame_width: number | null;
  frame_height: number | null;
  frames: TrackingOverlayFrame[];
}

export interface VideoProcessingResponse {
  summary: VideoSummary;
  reports: ViolationReport[];
  zone_violations?: ZoneViolation[];
  tracking_overlay?: TrackingOverlay;
}

export interface StreamEvent {
  event: "frame" | "violation" | "zone_violation" | "summary" | "error" | "start" | "end";
  frame_index?: number;
  data: any;
}
